<h1>Capstone 1: Data Wrangling (from CSV)</h1>

<a id='TOC'></a>
<strong>Table of Contents</strong>
<ol>
    <li>Preliminaries</li>
    <ol>
        <li><a href=#Sec01A>Import Modules</a></li>
    </ol>
    <li>Define Functions</li>
    <ol>
        <li><a href=#Sec02A>Remove Unnecessary Values</a></li>
        <li><a href=#Sec02B>Reduce Size of DataFrame</a></li>
        <li><a href=#Sec02C>Redesign the DataFrame</a></li>
    </ol>
    <li>Data Acquisition</li>
    <ol>
        <li><a href=#Sec03A>Inspect Sample of EMS Dataset</a></li>
        <li><a href=#Sec03B>Import Full EMS Dataset</a></li>
        <li><a href=#Sec03C>Import Geographical Dataset</a></li>
        <li><a href=#Sec03D>Inspect Raw, Merged Dataset</a></li>
    </ol>
    <li>Data Wrangling</li>
    <ol>
        <li><a href=#Sec04A>Apply Custom Functions</a></li>
        <li><a href=#Sec04B>Inspect Clean Dataset</a></li>
        <li><a href=#Sec04C>Export Clean Dataset</a></li>
        <li><a href=#Sec04D>Process Summary</a></li>
    </ol>
</ol>

<p>The goal of this project is to develop machine learning models that predict whether or not the outcome of an EMS incident will result in a fatality. This is a supervised, binary classification problem. Analyses will be performed on a collection of nearly 8 million records of documented incidents, which span the six year period from January 2013 through December 2018, and appropriate predictive models will be developed to achieve the primary objective. This dataset is robust and contains several feature variables, of mixed data types, that describe both various attributes of each incident as well as the responsive action taken by the FDNY. All of the aforementioned factors affect an individual’s survivability once a response is initiated.</p>

<p>Data wrangling will be performed on two datasets for this analysis. One dataset contains EMS incident data spanning a six-year period. The second dataset contains geographical information for all ZIP Code Tabulation Areas within the City of New York.</p>

***

<h2 style="text-transform: uppercase;">1. Preliminaries</h2>

<a id='Sec01A'></a>
<h4>1A: Import Modules</h4>

In [1]:
# Import packages and modules
import pandas as pd
import numpy as np
import datetime
import time

<p><a href=#TOC>TOC</a></p>

<hr>

<h2 style="text-transform: uppercase;">2. Define Functions</h2>

<a id='Sec02A'></a>
<h4>2A: Remove Unnecessary Values (drop_invalid and drop_immaterial)</h4>

For the purpose of this analysis, any observation with a missing value for `incident_disposition_code` must be omitted since the target variable is derived from this feature. In addition, observations that contain the following outliers, errors, or immaterial information must also be removed from the dataset:
<ul>
    <li>incidents created to transport a patient from one facility to another</li>
    <li>incidents where units were assigned to stand by in case they were needed</li>
    <li>incidents that pertain to special events</li>
    <li>incidents that were once closed but later reopened</li>
    <li>incidents with calculation errors for duration metrics</li>
    <li>features that contain redundant geographic information for incident</li>
</ul>

In [2]:
def drop_invalid(dfObj):
    # Drop all rows with missing value for select features
    dfObj.dropna(subset=['incident_disposition_code',
                         'zipcode',
                         'dispatch_response_seconds_qy',
                         'incident_travel_tm_seconds_qy',
                         'incident_response_seconds_qy'],
                 inplace=True)
    
    # Identify all columns that validate duration metrics
    list_of_validation_cols = [name for name in list(dfObj.columns)
                               if 'valid' in str(name)]
    
    # Drop all rows with invalid duration metrics
    for name in list_of_validation_cols:
        invalid_idx = dfObj[dfObj[name]=='N'].index
        dfObj.drop(invalid_idx, inplace=True)
    
    # Drop all rows where EMS were not dispatched
    no_disp_idx = dfObj[dfObj.dispatch_response_seconds_qy==0].index
    dfObj.drop(no_disp_idx, inplace=True)
    
    return dfObj

In [3]:
def drop_immaterial(dfObj):
    # Identify all columns with outlier event indicators
    list_of_indicator_cols = [name for name in list(dfObj.columns) 
                              if 'indicator' in str(name) and name !='held_indicator']
    
    # Drop all rows that pertain to outlier incidents
    for name in list_of_indicator_cols:
        outlier_idx = dfObj[dfObj[name]=='Y'].index
        dfObj.drop(outlier_idx, inplace=True)
    
    # Remove columns that contain incident indicator data
    dfObj.drop(list_of_indicator_cols,axis=1,inplace=True)
    dfObj.drop([name for name in list(dfObj.columns) 
                if '_indc' in str(name)],axis=1,inplace=True)
    
    # Identify and remove all columns that contain redundant geographic data
    list_of_zone_cols = [name for name in list(dfObj.columns)
                         if (('district' in str(name))| 
                             (name =='policeprecinct')| 
                             (name =='geoid'))]
    dfObj.drop(list_of_zone_cols,axis=1,inplace=True)
    
    return dfObj

<p><a href=#TOC>TOC</a></p>

<a id='Sec02B'></a>
<h4>2B: Reduce Size of DataFrame (reduce_memory)</h4>

<p>Modifying the data types for values contained within select columns will drastically reduce the memory usage of the dataframe object.</p>

In [4]:
def reduce_memory(dfObj):
    # Truncate name for borough label: 'RICHMOND / STATEN ISLAND'
    dfObj['borough'] = dfObj.borough.replace('RICHMOND / STATEN ISLAND',
                                             'STATEN ISLAND')
    
    # Create list of all columns that contain ISO8601 datetime
    list_of_datetime_cols = [name for name in list(dfObj.columns) 
                             if 'datetime' in str(name)]

    # Convert dtypes for each element in list to datetime
    for name in list_of_datetime_cols:
        dfObj[name] = pd.to_datetime(dfObj[name],errors='coerce')
       
    # Create list of all columns that contain time duration
    list_of_numeric_cols = [name for name in list(dfObj.columns) 
                            if (('seconds' in str(name))|
                                ('severity' in str(name))|
                                ('disposition' in str(name))|
                                ('cad' in str(name)))]

    # Convert dtypes for each element in list to numeric
    for name in list_of_numeric_cols:
        dfObj[name] = pd.to_numeric(dfObj[name],errors='coerce')
        
    # Convert columns to category dtypes to reduce size of dataframe object
    dfObj['borough'] = dfObj.borough.astype('category')
    dfObj['zipcode'] = dfObj.zipcode.astype('category')
    dfObj['held_indicator'] = dfObj.held_indicator.astype('category')
    dfObj['incident_dispatch_area'] = dfObj.incident_dispatch_area.astype('category')
    
    return dfObj

<p><a href=#TOC>TOC</a></p>

<a id='Sec02C'></a>
<h4>2C: Redesign the DataFrame (format_df)</h4>

Construct a boolean series that represents the target variable (fatality) using the corresponding values in `incident_disposition_code`. Also, apply aesthetic changes to help improve the readability of the dataframe object.

In [5]:
def format_df(dfObj):
    # Rename select columns
    dfObj.rename(columns={'initial_severity_level_code':'initial_severity_level',
                          'final_severity_level_code':'final_severity_level',
                          'dispatch_response_seconds_qy':'dispatch_time',
                          'incident_travel_tm_seconds_qy':'travel_time',
                          'incident_response_seconds_qy':'response_time',
                          'intptlat':'latitude',
                          'intptlong':'longitude'},inplace=True)
    
    # Create a series for a new feature variable: life_threatening
    dfObj['life_threatening'] = [True if ((val ==1)|
                                         (val == 2)|
                                         (val == 3)) else False 
                                 for val in dfObj['final_severity_level'].astype('int64')]
    
    # Create a series for the target variable: fatality
    dfObj['fatality'] = np.logical_or(dfObj.incident_disposition_code.astype('int64') == 83,
                                      dfObj.incident_disposition_code.astype('int64') == 96)
    
    # Create separate columns for time components of the incident
    dfObj['year'] = pd.DatetimeIndex(dfObj.incident_datetime).strftime('%Y')
    dfObj['month'] = pd.DatetimeIndex(dfObj.incident_datetime).strftime('%m')
    dfObj['day'] = pd.DatetimeIndex(dfObj.incident_datetime).strftime('%d')
    dfObj['weekday'] = pd.DatetimeIndex(dfObj.incident_datetime).strftime('%w')
    dfObj['hour'] = pd.DatetimeIndex(dfObj.incident_datetime).strftime('%H')
    
    # Convert dtypes for each element in list to numeric
    list_of_time_cols = ['year','month','day','weekday','hour']
    for name in list_of_time_cols:
        dfObj[name] = pd.to_numeric(dfObj[name],errors='coerce')
    
    # Reorder columns of DataFrame object
    col_order = ['cad_incident_id','incident_datetime',
                 'year','month','day','hour','weekday','borough',
                 'zipcode','latitude','longitude','aland_sqmi','awater_sqmi',
                 'initial_call_type','initial_severity_level',
                 'final_call_type','final_severity_level',
                 'held_indicator','first_assignment_datetime',
                 'incident_dispatch_area',
                 'dispatch_time','first_activation_datetime',
                 'first_on_scene_datetime','travel_time','response_time',
                 'first_to_hosp_datetime','first_hosp_arrival_datetime',
                 'incident_close_datetime','incident_disposition_code',
                 'life_threatening','fatality']
    dfObj=dfObj[col_order]
    
    return dfObj

<p><a href=#TOC>TOC</a></p>

<hr>

<h2 style="text-transform: uppercase;">3. Data Acquisition</h2>

<a id='Sec03A'></a>
<h4>3A: Inspect Sample of EMS Incident Dataset</h4>

In [6]:
# Import sample of EMS incident dataset into a Pandas DataFrame
"""
    The original CSV file can be exported from 
    https://data.cityofnewyork.us/Public-Safety/EMS-Incident-Dispatch-Data/76xm-jjuj
"""
input_path_ems = '../data/EMS_Incident_Dispatch_Data.csv'
preview_df = pd.read_csv(input_path_ems,header=0,nrows=1000)

In [7]:
preview_df.shape

(1000, 32)

In [8]:
preview_df.head()

CAD_INCIDENT_ID       INCIDENT_DATETIME INITIAL_CALL_TYPE  \
0        130010001  01/01/2013 12:00:04 AM            RESPIR   
1        130010002  01/01/2013 12:00:19 AM              CARD   
2        130010004  01/01/2013 12:01:04 AM            ARREST   
3        130010005  01/01/2013 12:01:16 AM              SICK   
4        130010006  01/01/2013 12:01:26 AM            INJURY   

   INITIAL_SEVERITY_LEVEL_CODE FINAL_CALL_TYPE  FINAL_SEVERITY_LEVEL_CODE  \
0                            4          RESPIR                          4   
1                            3            CARD                          3   
2                            1          ARREST                          1   
3                            6            SICK                          6   
4                            5          INJURY                          5   

  FIRST_ASSIGNMENT_DATETIME VALID_DISPATCH_RSPNS_TIME_INDC  \
0    01/01/2013 12:01:45 AM                              Y   
1    01/01/2013 12:01:18 AM                              Y   
2    01/01/2013 12:01:33 AM                              Y   
3    01/01/2013 12:02:12 AM                              Y   
4    01/01/2013 12:01:58 AM                              Y   

   DISPATCH_RESPONSE_SECONDS_QY FIRST_ACTIVATION_DATETIME        ...          \
0                           101    01/01/2013 12:01:51 AM        ...           
1                            59    01/01/2013 12:02:08 AM        ...           
2                            29    01/01/2013 12:01:58 AM        ...           
3                            56    01/01/2013 12:02:55 AM        ...           
4                            32    01/01/2013 12:02:55 AM        ...           

   ZIPCODE POLICEPRECINCT  CITYCOUNCILDISTRICT  COMMUNITYDISTRICT  \
0  10472.0           43.0                 18.0              209.0   
1  10454.0           40.0                  8.0              201.0   
2  11418.0          102.0                 29.0              409.0   
3  10453.0           46.0                 14.0              205.0   
4  10457.0           48.0                 15.0              206.0   

  COMMUNITYSCHOOLDISTRICT CONGRESSIONALDISTRICT REOPEN_INDICATOR  \
0                    12.0                  15.0                N   
1                     7.0                  15.0                N   
2                    27.0                   5.0                N   
3                    10.0                  15.0                N   
4                    10.0                  15.0                N   

  SPECIAL_EVENT_INDICATOR  STANDBY_INDICATOR TRANSFER_INDICATOR  
0                       N                  N                  N  
1                       N                  N                  N  
2                       N                  N                  N  
3                       N                  N                  N  
4                       N                  N                  N  

[5 rows x 32 columns]

In [9]:
preview_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 32 columns):
CAD_INCIDENT_ID                   1000 non-null int64
INCIDENT_DATETIME                 1000 non-null object
INITIAL_CALL_TYPE                 1000 non-null object
INITIAL_SEVERITY_LEVEL_CODE       1000 non-null int64
FINAL_CALL_TYPE                   1000 non-null object
FINAL_SEVERITY_LEVEL_CODE         1000 non-null int64
FIRST_ASSIGNMENT_DATETIME         935 non-null object
VALID_DISPATCH_RSPNS_TIME_INDC    1000 non-null object
DISPATCH_RESPONSE_SECONDS_QY      1000 non-null int64
FIRST_ACTIVATION_DATETIME         932 non-null object
FIRST_ON_SCENE_DATETIME           890 non-null object
VALID_INCIDENT_RSPNS_TIME_INDC    1000 non-null object
INCIDENT_RESPONSE_SECONDS_QY      889 non-null float64
INCIDENT_TRAVEL_TM_SECONDS_QY     890 non-null float64
FIRST_TO_HOSP_DATETIME            522 non-null object
FIRST_HOSP_ARRIVAL_DATETIME       513 non-null object
INCIDENT_CLOSE_DATETIME

<p><a href=#TOC>TOC</a></p>

<a id='Sec03B'></a>
<h4>3B: Import Full EMS Incident Dataset (from CSV file)</h4>

In [10]:
# Import full EMS incident dataset into a Pandas DataFrame object
"""
    The original CSV file can be exported from 
    https://data.cityofnewyork.us/Public-Safety/EMS-Incident-Dispatch-Data/76xm-jjuj
"""
input_path_ems = '../data/EMS_Incident_Dispatch_Data.csv'
ems_df = pd.read_csv(input_path_ems,header=0)

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2785: DtypeWarning: Columns (14,15,20) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [11]:
# Convert all column header names to lowercase
ems_df.rename(columns={i:i.lower().strip() for i in ems_df.columns},inplace=True)
ems_df.columns

Index(['cad_incident_id', 'incident_datetime', 'initial_call_type',
       'initial_severity_level_code', 'final_call_type',
       'final_severity_level_code', 'first_assignment_datetime',
       'valid_dispatch_rspns_time_indc', 'dispatch_response_seconds_qy',
       'first_activation_datetime', 'first_on_scene_datetime',
       'valid_incident_rspns_time_indc', 'incident_response_seconds_qy',
       'incident_travel_tm_seconds_qy', 'first_to_hosp_datetime',
       'first_hosp_arrival_datetime', 'incident_close_datetime',
       'held_indicator', 'incident_disposition_code', 'borough', 'atom',
       'incident_dispatch_area', 'zipcode', 'policeprecinct',
       'citycouncildistrict', 'communitydistrict', 'communityschooldistrict',
       'congressionaldistrict', 'reopen_indicator', 'special_event_indicator',
       'standby_indicator', 'transfer_indicator'],
      dtype='object')

In [12]:
# Obtain shape of DataFrame w/ EMS incident data
ems_df.shape

(8557848, 32)

<p>A downloadable description of each dataset field is available at <a href="https://data.cityofnewyork.us/Public-Safety/EMS-Incident-Dispatch-Data/76xm-jjuj" target="_blank">NYC Open Data</a> in the <em>Attachments</em> section under the file name <strong>EMS_incident_dispatch_data_description.xlsx</strong>. <a href=#TOC>TOC</a></p>

<a id='Sec03C'></a>
<h4>3C: Import Geographical Dataset (from tab-delimited file)</h4>

In [13]:
# Import geographic data for all U.S. cities into a Pandas DataFrame object
"""
    The original TXT file can be exported from
    https://www2.census.gov/geo/docs/maps-data/data/gazetteer/2019_Gazetteer/2019_Gaz_zcta_national.zip
"""
input_path_geo = '../data/2019_Gaz_zcta_national.txt'
geo_df = pd.read_table(input_path_geo)

In [14]:
# Convert all column header names to lowercase
geo_df.rename(columns={i:i.lower().strip() for i in geo_df.columns},inplace=True)
geo_df.columns

Index(['geoid', 'aland', 'awater', 'aland_sqmi', 'awater_sqmi', 'intptlat',
       'intptlong'],
      dtype='object')

In [15]:
# Obtain shape of DataFrame w/ geographic data
geo_df.shape

(33144, 7)

<p>The text file contains GPA coordinates for all ZIP Code Tabulation Areas (ZCTAs) within the United States. A downloadable file is available at the <a href="https://www.census.gov/geographies/reference-files/time-series/geo/gazetteer-files.html" target="_blank">U.S. Census Bureau</a> under the file name <strong>ZIP Code Tabulation Areas</strong>. <a href=#TOC>TOC</a></p>

<a id='Sec03D'></a>
<h4>3D: Inspect Raw, Merged Dataset</h4>

In [16]:
# Join the EMS incidents and geographic datasets
df = pd.merge(ems_df,geo_df,left_on='zipcode',right_on='geoid')

In [17]:
df.shape

(8368586, 39)

In [18]:
df.head()

cad_incident_id       incident_datetime initial_call_type  \
0        130010001  01/01/2013 12:00:04 AM            RESPIR   
1        130010022  01/01/2013 12:05:52 AM               EDP   
2        130010086  01/01/2013 12:20:37 AM              SICK   
3        130010615  01/01/2013 01:53:11 AM            INJURY   
4        130010624  01/01/2013 01:54:28 AM              SICK   

   initial_severity_level_code final_call_type  final_severity_level_code  \
0                            4          RESPIR                          4   
1                            7             EDP                          7   
2                            6            SICK                          6   
3                            4          INJURY                          4   
4                            4            SICK                          4   

  first_assignment_datetime valid_dispatch_rspns_time_indc  \
0    01/01/2013 12:01:45 AM                              Y   
1    01/01/2013 12:06:04 AM                              Y   
2    01/01/2013 12:21:29 AM                              Y   
3    01/01/2013 01:53:36 AM                              Y   
4    01/01/2013 01:55:10 AM                              Y   

   dispatch_response_seconds_qy first_activation_datetime    ...     \
0                           101    01/01/2013 12:01:51 AM    ...      
1                            12    01/01/2013 12:06:24 AM    ...      
2                            52    01/01/2013 12:21:50 AM    ...      
3                            25    01/01/2013 01:53:52 AM    ...      
4                            42    01/01/2013 01:55:32 AM    ...      

  special_event_indicator standby_indicator  transfer_indicator  geoid  \
0                       N                 N                   N  10472   
1                       N                 N                   N  10472   
2                       N                 N                   N  10472   
3                       N                 N                   N  10472   
4                       N                 N                   N  10472   

     aland awater aland_sqmi awater_sqmi   intptlat intptlong  
0  2729341      0      1.054         0.0  40.829556 -73.86931  
1  2729341      0      1.054         0.0  40.829556 -73.86931  
2  2729341      0      1.054         0.0  40.829556 -73.86931  
3  2729341      0      1.054         0.0  40.829556 -73.86931  
4  2729341      0      1.054         0.0  40.829556 -73.86931  

[5 rows x 39 columns]

In [19]:
df.info(verbose=True,null_counts=True,memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8368586 entries, 0 to 8368585
Data columns (total 39 columns):
cad_incident_id                   8368586 non-null int64
incident_datetime                 8368586 non-null object
initial_call_type                 8368586 non-null object
initial_severity_level_code       8368586 non-null int64
final_call_type                   8368586 non-null object
final_severity_level_code         8368586 non-null int64
first_assignment_datetime         8312407 non-null object
valid_dispatch_rspns_time_indc    8368586 non-null object
dispatch_response_seconds_qy      8368586 non-null int64
first_activation_datetime         8297820 non-null object
first_on_scene_datetime           8106919 non-null object
valid_incident_rspns_time_indc    8368586 non-null object
incident_response_seconds_qy      8105959 non-null float64
incident_travel_tm_seconds_qy     8106798 non-null float64
first_to_hosp_datetime            5935067 non-null object
first_hosp_arrival_

In [20]:
df.memory_usage(deep=True)

Index                              66948688
cad_incident_id                    66948688
incident_datetime                 661118294
initial_call_type                 518136144
initial_severity_level_code        66948688
final_call_type                   518235008
final_severity_level_code          66948688
first_assignment_datetime         658477881
valid_dispatch_rspns_time_indc    518852332
dispatch_response_seconds_qy       66948688
first_activation_datetime         657792292
first_on_scene_datetime           648819945
valid_incident_rspns_time_indc    518852332
incident_response_seconds_qy       66948688
incident_travel_tm_seconds_qy      66948688
first_to_hosp_datetime            546742901
first_hosp_arrival_datetime       544847767
incident_close_datetime           660970902
held_indicator                    518852332
incident_disposition_code          66948688
borough                           542579701
atom                              468511012
incident_dispatch_area          

<p><a href=#TOC>TOC</a></p>

<hr>

<h2 style="text-transform: uppercase;">4. Data Wrangling</h2>

<a id='Sec04A'></a>
<h4>4A: Apply Custom Functions to Dataset</h4>

In [21]:
# Apply drop_invalid to DataFrame object
start_time = time.time()
print('Applying "drop_invalid" function...')
df = drop_invalid(df)
print('Process complete.')
print(f'Runtime: {time.time()-start_time:.4f} seconds')

Applying "drop_invalid" function...
Process complete.
Runtime: 163.7182 seconds


In [22]:
# Apply drop_immaterial to DataFrame object
start_time = time.time()
print('Applying "drop_immaterial" function...')
df = drop_immaterial(df)
print('Process complete.')
print(f'Runtime: {time.time()-start_time:.4f} seconds')

Applying "drop_immaterial" function...
Process complete.
Runtime: 298.9684 seconds


In [23]:
# Apply reduce_memory to DataFrame object
start_time = time.time()
print('Applying "reduce_memory" function...')
df = reduce_memory(df)
print('Process complete.')
print(f'Runtime: {time.time()-start_time:.4f} seconds')

Applying "reduce_memory" function...
Process complete.
Runtime: 6546.2373 seconds


In [24]:
# Apply format_df to DataFrame object
start_time = time.time()
print('Applying "format_df" function...')
df = format_df(df)
print('Process complete.')
print(f'Runtime: {time.time()-start_time:.4f} seconds')

Applying "format_df" function...
Process complete.
Runtime: 224.7236 seconds


In [25]:
# Set index to 'cad_incident_id'
df.set_index(['cad_incident_id'],inplace=True)

<p><a href=#TOC>TOC</a></p>

<a id='Sec04B'></a>
<h4>4B: Inspect Clean Dataset</h4>

In [26]:
df.shape

(7988028, 30)

In [27]:
df.head()

incident_datetime  year  month  day  hour  weekday borough  \
cad_incident_id                                                                
130010001       2013-01-01 00:00:04  2013      1    1     0        2   BRONX   
130010022       2013-01-01 00:05:52  2013      1    1     0        2   BRONX   
130010086       2013-01-01 00:20:37  2013      1    1     0        2   BRONX   
130010615       2013-01-01 01:53:11  2013      1    1     1        2   BRONX   
130010624       2013-01-01 01:54:28  2013      1    1     1        2   BRONX   

                 zipcode   latitude  longitude    ...     \
cad_incident_id                                   ...      
130010001        10472.0  40.829556  -73.86931    ...      
130010022        10472.0  40.829556  -73.86931    ...      
130010086        10472.0  40.829556  -73.86931    ...      
130010615        10472.0  40.829556  -73.86931    ...      
130010624        10472.0  40.829556  -73.86931    ...      

                 first_activation_datetime  first_on_scene_datetime  \
cad_incident_id                                                       
130010001              2013-01-01 00:01:51      2013-01-01 00:13:21   
130010022              2013-01-01 00:06:24      2013-01-01 00:14:46   
130010086              2013-01-01 00:21:50      2013-01-01 00:32:14   
130010615              2013-01-01 01:53:52      2013-01-01 01:56:54   
130010624              2013-01-01 01:55:32      2013-01-01 01:59:26   

                travel_time  response_time first_to_hosp_datetime  \
cad_incident_id                                                     
130010001             696.0          797.0    2013-01-01 00:28:49   
130010022             522.0          534.0    2013-01-01 00:48:57   
130010086             645.0          697.0                    NaT   
130010615             198.0          223.0    2013-01-01 02:12:28   
130010624             256.0          298.0    2013-01-01 02:14:34   

                 first_hosp_arrival_datetime incident_close_datetime  \
cad_incident_id                                                        
130010001                2013-01-01 00:38:15     2013-01-01 01:04:56   
130010022                2013-01-01 01:02:02     2013-01-01 01:46:14   
130010086                                NaT     2013-01-01 01:03:50   
130010615                2013-01-01 02:26:09     2013-01-01 03:03:36   
130010624                2013-01-01 02:23:06     2013-01-01 02:44:27   

                incident_disposition_code life_threatening  fatality  
cad_incident_id                                                       
130010001                            82.0            False     False  
130010022                            82.0            False     False  
130010086                            93.0            False     False  
130010615                            82.0            False     False  
130010624                            82.0            False     False  

[5 rows x 30 columns]

In [28]:
df.info(verbose=True,null_counts=True,memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7988028 entries, 130010001 to 183530054
Data columns (total 30 columns):
incident_datetime              7988028 non-null datetime64[ns]
year                           7988028 non-null int64
month                          7988028 non-null int64
day                            7988028 non-null int64
hour                           7988028 non-null int64
weekday                        7988028 non-null int64
borough                        7988028 non-null category
zipcode                        7988028 non-null category
latitude                       7988028 non-null float64
longitude                      7988028 non-null float64
aland_sqmi                     7988028 non-null float64
awater_sqmi                    7988028 non-null float64
initial_call_type              7988028 non-null object
initial_severity_level         7988028 non-null int64
final_call_type                7988028 non-null object
final_severity_level           7988028 non

In [29]:
df.memory_usage(deep=True)

Index                           63904224
incident_datetime               63904224
year                            63904224
month                           63904224
day                             63904224
hour                            63904224
weekday                         63904224
borough                          7988578
zipcode                         15987960
latitude                        63904224
longitude                       63904224
aland_sqmi                      63904224
awater_sqmi                     63904224
initial_call_type              494612046
initial_severity_level          63904224
final_call_type                494709109
final_severity_level            63904224
held_indicator                   7988232
first_assignment_datetime       63904224
incident_dispatch_area           7991432
dispatch_time                   63904224
first_activation_datetime       63904224
first_on_scene_datetime         63904224
travel_time                     63904224
response_time   

<p><a href=#TOC>TOC</a></p>

<a id='Sec04C'></a>
<h4>4C: Export Clean Dataset (to CSV)</h4>

In [30]:
output_path = '../data/clean_EMS_data_from_csv.csv'
print('Exporting DataFrame to CSV...')
df.to_csv(output_path,index=False,compression='gzip')
print('DataFrame successfully exported to CSV using \'gzip\' compression.')

Exporting DataFrame to CSV...
DataFrame successfully exported to CSV using 'gzip' compression.


<p><a href=#TOC>TOC</a></p>

<a id='Sec04D'></a>
<h4>4D: Process Summary</h4>

<p>The original EMS incident dataset was comprised of 8,557,848 observations with 32 variables of mixed data types, and its source file occupied approximately 1.98 GB of hard disk space. The original geographic info dataset was comprised of 33,144 observations with 7 variables, and its source file occupied 6.7 MB in system memory. The DataFrame objects generated from both source files were joined on the 'ZIP code' feature, though the labels had different names in each object.</p>

The resulting raw dataset consists of 8,368,586 observations with 39 variables, and occupies 11.1 GB in system memory. The target variable (`fatality`) was created by applying a boolean filter on the `incident_disposition_code` column within the DataFrame, which indicates the outcome of any EMS incident.

<p>After all data pre-processing is complete, the resulting clean dataset consists of 7,988,028 observations of mixed data types, with a clear target variable and 26 feature variables. Its output file occupies 331 MB of hard disk space and 2.3 GB in system memory. <a href=#TOC>TOC</a></p>